lib/execution_segment.go

```diff
@@ -259,27 +260,25 @@ func roundUp(rat *big.Rat) *big.Int {
 
 // Scale proportionally scales the supplied value, according to the execution
 // segment's position and size of the work.
+// This implements the first proposal from #997:
+//   floor( (value * from) % 1 + value * length )
 func (es *ExecutionSegment) Scale(value int64) int64 {
 	if es == nil { // no execution segment, i.e. 100%
 		return value
 	}
-	// Instead of the first proposal that used remainders and floor:
-	//    floor( (value * from) % 1 + value * length )
-	// We're using an alternative approach with rounding that (hopefully) has
-	// the same properties, but it's simpler and has better precision:
-	//    round( (value * from) - round(value * from) + (value * (to - from)) )?
-	// which reduces to:
-	//    round( (value * to) - round(value * from) )?
-
-	toValue := big.NewRat(value, 1)
-	toValue.Mul(toValue, es.to)
-
 	fromValue := big.NewRat(value, 1)
 	fromValue.Mul(fromValue, es.from)
 
-	toValue.Sub(toValue, new(big.Rat).SetFrac(roundUp(fromValue), oneBigInt))
+	rem := new(big.Int).Rem(fromValue.Num(), fromValue.Denom())
+	fromRem := big.NewRat(rem.Int64(), fromValue.Denom().Int64())
+
+	lenValue := big.NewRat(value, 1)
+	lenValue.Mul(lenValue, es.length)
+
+	total := new(big.Rat).Add(fromRem, lenValue)
```

Comment:
Given that both `lenValue` and `fromRem` aren't going to be used for anything else I would argue you can put the result in one of them as this saves you some allocation.
I also recommend to write a simple benchmark while doign optimizations :D

Reply:
Thanks, fixed in latest push.

Also added a benchmark, even though this wasn't an optimization effort, but more of a functional one to compare both implementations.

FWIW, here's the benchstat output comparing current `new-schedulers` and this new "old algorithm" implementation:

```
name                      old time/op    new time/op    delta
ExecutionSegmentScale-12    1.29µs ±16%    1.50µs ± 9%  +15.81%  (p=0.000 n=20+19)

name                      old alloc/op   new alloc/op   delta
ExecutionSegmentScale-12      488B ± 0%      480B ± 0%   -1.64%  (p=0.000 n=20+20)

name                      old allocs/op  new allocs/op  delta
ExecutionSegmentScale-12      25.0 ± 0%      25.0 ± 0%     ~     (all equal)
```

So a bit slower, but I don't think it should be a concern.